flink-libraries/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliChangelogResultView.java

```diff
@@ -133,6 +134,13 @@ protected void refresh() {
 					}
 
 					// update results
+
+					// formatting and printing of rows is expensive in the current implementation,
+					// therefore we limit the maximum number of lines shown in changelog mode to
+					// keep the CLI responsive
+					if (results.size() >= DEFAULT_MAX_ROW_COUNT) {
+						results.remove(0);
```

Comment:
Since `results` is an `ArrayList`, removing the first element means that the whole array is shifted by copying. `LinkedList` might be more efficient.

Reply:
Alternatively, we could delete all entries at once with `ArrayList.removeRange()`